File: next/Tools/ReviewTools.fs
namespace Wanxiangshu.Next.Tools

open System.Threading.Tasks
open Thoth.Json
open Wanxiangshu.Next.Session

module ReviewTools =

    let submitReviewTool (port: SessionCommandPort) : Tool =
        { Name = "submit_review"
          Description = "Submit review task result."
          SchemaJson = """{"type":"object","properties":{"report":{"type":"string"}},"required":["report"]}"""
          Execute =
            fun ctx input ->
                task {
                    ctx.Cancellation.ThrowIfCancellationRequested()

                    let reportText =
                        try
                            let decoder = Decode.field "report" Decode.string

                            match Decode.fromString decoder input.Payload with
                            | Ok r -> r
                            | Error _ -> input.Payload
                        with _ ->
                            input.Payload

                    let replyRef = ref (Ok SessionCommandResult.ReviewSubmitted)
                    let cmd = SubmitReview(reportText, (fun r -> replyRef := r))
                    let! res = port.Request cmd ctx.Cancellation ctx.Deadline

                    match res with
                    | Ok SessionCommandResult.ReviewSubmitted ->
                        return
                            { Result = "Review submitted and structured fact recorded"
                              Truncated = false }
                    | Ok _ ->
                        return
                            { Result = "Review submitted"
                              Truncated = false }
                    | Error err ->
                        return
                            { Result = sprintf "Failed to record review submission: %A" err
                              Truncated = false }
                } }

    let returnReviewerTool (port: SessionCommandPort) : Tool =
        { Name = "return_reviewer"
          Description = "Return verdict from reviewer."
          SchemaJson = """{"type":"object","properties":{"verdict":{"type":"string"}},"required":["verdict"]}"""
          Execute =
            fun ctx input ->
                task {
                    ctx.Cancellation.ThrowIfCancellationRequested()

                    let verdictText =
                        try
                            let decoder = Decode.field "verdict" Decode.string

                            match Decode.fromString decoder input.Payload with
                            | Ok v -> v
                            | Error _ -> input.Payload
                        with _ ->
                            input.Payload

                    let replyRef = ref (Ok SessionCommandResult.VerdictReturned)
                    let cmd = ReturnVerdict(verdictText, (fun r -> replyRef := r))
                    let! res = port.Request cmd ctx.Cancellation ctx.Deadline

                    match res with
                    | Ok SessionCommandResult.VerdictReturned ->
                        return
                            { Result = sprintf "Reviewer verdict returned: %s" verdictText
                              Truncated = false }
                    | Ok _ ->
                        return
                            { Result = sprintf "Reviewer verdict returned: %s" verdictText
                              Truncated = false }
                    | Error err ->
                        return
                            { Result = sprintf "Failed to record reviewer verdict: %A" err
                              Truncated = false }
                } }

File: .wanxiangshu/runtimes/Tools/ReviewTools.js

import { task } from "../fable_modules/fable-library-js.5.6.0/TaskBuilder.js";
import { throwIfCancellationRequested } from "../fable_modules/fable-library-js.5.6.0/Async.js";
import { string, field, fromString } from "../fable_modules/Thoth.Json.10.5.1/Decode.fs.js";
import { SessionCommand, SessionCommandResult } from "../Session/Inbox.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.5.6.0/Result.js";
import { printf, toText } from "../fable_modules/fable-library-js.5.6.0/String.js";
import { Tool, ToolOutput } from "./ToolContext.js";

export function submitReviewTool() {
    return new Tool("submit_review", "Submit review task result.", "{\"type\":\"object\",\"properties\":{\"report\":{\"type\":\"string\"}},\"required\":[\"report\"]}", (ctx, input) => {
        const builder$0040 = task();
        return builder$0040.Run(builder$0040.Delay(() => {
            throwIfCancellationRequested(ctx.Cancellation);
            let reportText;
            try {
                const matchValue = fromString((path_1, value_1) => field("report", string, path_1, value_1), input.Payload);
                reportText = ((matchValue.tag === 1) ? input.Payload : matchValue.fields[0]);
            }
            catch (matchValue_1) {
                reportText = input.Payload;
            }
            let replyVal = new FSharpResult$2(0, [new SessionCommandResult(2, [])]);
            return builder$0040.Bind(ctx.Session.Request(new SessionCommand(2, [reportText, (r_1) => {
                replyVal = r_1;
            }]), ctx.Cancellation, ctx.Deadline), (_arg) => {
                const res = _arg;
                return (res.tag === 1) ? builder$0040.Return(new ToolOutput(toText(printf("Failed to record review submission: %A"))(res.fields[0]), false)) : ((res.fields[0].tag === 2) ? builder$0040.Return(new ToolOutput("Review submitted and structured fact recorded", false)) : builder$0040.Return(new ToolOutput("Review submitted", false)));
            });
        }));
    });
}

export function returnReviewerTool() {
    return new Tool("return_reviewer", "Return verdict from reviewer.", "{\"type\":\"object\",\"properties\":{\"verdict\":{\"type\":\"string\"}},\"required\":[\"verdict\"]}", (ctx, input) => {
        const builder$0040 = task();
        return builder$0040.Run(builder$0040.Delay(() => {
            throwIfCancellationRequested(ctx.Cancellation);
            let verdictText;
            try {
                const matchValue = fromString((path_1, value_1) => field("verdict", string, path_1, value_1), input.Payload);
                verdictText = ((matchValue.tag === 1) ? input.Payload : matchValue.fields[0]);
            }
            catch (matchValue_1) {
                verdictText = input.Payload;
            }
            let replyVal = new FSharpResult$2(0, [new SessionCommandResult(3, [])]);
            return builder$0040.Bind(ctx.Session.Request(new SessionCommand(3, [verdictText, (r) => {
                replyVal = r;
            }]), ctx.Cancellation, ctx.Deadline), (_arg) => {
                const res = _arg;
                return (res.tag === 1) ? builder$0040.Return(new ToolOutput(toText(printf("Failed to record reviewer verdict: %A"))(res.fields[0]), false)) : ((res.fields[0].tag === 3) ? builder$0040.Return(new ToolOutput(toText(printf("Reviewer verdict returned: %s"))(verdictText), false)) : builder$0040.Return(new ToolOutput(toText(printf("Reviewer verdict returned: %s"))(verdictText), false)));
            });
        }));
    });
}

